core/src/main/java/org/apache/iceberg/rest/RESTClient.java

```diff
@@ -27,9 +27,14 @@
  * Interface for a basic HTTP Client for interfacing with the REST catalog.
  */
 public interface RESTClient extends Closeable {
-  <T> T delete(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
-  <T> T post(String path, Object body, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
-  <T> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
-  void  head(String path, Consumer<ErrorResponse> errorHandler);
+
+  void head(String path, Consumer<ErrorResponse> errorHandler);
+
+  <T extends RESTResponse<T>> T delete(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
+
+  <T extends RESTResponse<T>> T get(String path, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
+
+  <T extends RESTResponse<T>> T post(
+      String path, Object body, Class<T> responseType, Consumer<ErrorResponse> errorHandler);
```

Comment:
Style wise, I wasn't really sure the best way to deal with these. Specifically because `post` would require multiple lines for the arguments.

I moved `head` up because it returns `void` now. 

Please feel free to make suggestions @rdblue